### DISCUSSION TITLE
Exit code when running multiple repositories

### DISCUSSION THREAD

**Discussion by chfern** [renovatebot/renovate#24307] [Request Help] [answered]
### How are you running Renovate?

Self hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.138.1

### If you're self-hosting Renovate, select which platform you are using.

Gitlab self hosted

### Was this something which used to work for you, and then stopped?

Placeholder value, please select the correct response from the dropdown

### Wanted end result.

I don't have any expected end result as of now. Just wanted to clarify the behavior.

Let's say we have 2 repositories. And if we run the repositories individually it will give result as such:
- repository A: renovate does not have push access to this repo, hence should error and exit with non-zero code
- repository B: renovate able to finish successfully. Exit code 0

However, if we run renovate passing the two repositories together (e.g with env var `RENOVATE_REPOSITORIES=repositoryA,repositoryB`), I found out that the exit code that renovate command returns depends on the last repository. In this case, it will return successfully even though repositoryA failed.

If this the expected behavior? Thanks before

### What you tried so far.

- Search whether there's any relevant config that may affect this
- Search for similar issue (but can't find any)

### Relevant debug logs

<details><summary>Logs</summary>

Run two repositories (firstRepo and secondRepo -- these are placeholder names). firstRepo failed and secondRepo succeeded. The exit code is 0 even though firstRepo failed.

```
Using docker image sha256:b307c2dbc2e091cce0004bcfb72de33d18361b772da2bc41b3fd8413234d2358 for renovate/renovate:35.138.1-slim with digest ...
$ echo "Running renovate."
Running renovate.
$ set +e
$ exit_code=0
$ LOG_FORMAT=json RENOVATE_REPOSITORIES=username/firstRepo,username/secondRepo renovate || exit_code=$?
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/firstRepo","renovateVersion":"35.138.1","msg":"Repository started","time":"2023-09-04T08:00:41.700Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/firstRepo","baseBranch":"master","stats":{"managers":{"gomod":{"fileCount":1,"depCount":21}},"total":{"fileCount":1,"depCount":21}},"msg":"Dependency extraction complete","time":"2023-09-04T08:01:19.326Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":40,"logContext":"omitted","repository":"username/firstRepo","branch":"renovate-0-go-github.com/gin-gonic/gin-vulnerability","err":{"task":{"commands":["push","origin","renovate-0-go-github.com/gin-gonic/gin-vulnerability:renovate-0-go-github.com/gin-gonic/gin-vulnerability","--force-with-lease","-u","--no-verify","--verbose","--porcelain"],"format":"utf-8","parser":"[function]"},"message":"Pushing to https://gitlabhost/username/firstRepo.git\nremote: You are not allowed to push code to this project.\nfatal: unable to access 'https://gitlabhost/username/firstRepo.git/': The requested URL returned error: 403\n","stack":"Error: Pushing to https://gitlabhost/username/firstRepo.git\nremote: You are not allowed to push code to this project.\nfatal: unable to access 'https://gitlabhost/username/firstRepo.git/': The requested URL returned error: 403\n\n    at Object.action (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"},"msg":"Error updating branch","time":"2023-09-04T08:02:46.698Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/firstRepo","cloned":true,"durationMs":129111,"msg":"Repository finished","time":"2023-09-04T08:02:50.811Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/secondRepo","renovateVersion":"35.138.1","msg":"Repository started","time":"2023-09-04T08:02:50.820Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/secondRepo","msg":"Repository has no package files - skipping","time":"2023-09-04T08:03:16.010Z","v":0}
{"name":"renovate","hostname":"omitted","pid":15,"level":30,"logContext":"omitted","repository":"username/secondRepo","cloned":true,"durationMs":25194,"msg":"Repository finished","time":"2023-09-04T08:03:16.012Z","v":0}
$ echo $exit_code
0
```

</details>

**Comment by rarkins** [comment, marked as answer]
Renovate should edit with a non zero code if any errors are logged with any repo

**Reply by chfern** [reply]
ahh, so it is because `firstRepo`'s error is logged with level 40 (which is warning).. I tried another repo that has a different error (logged with level 50) now it exits early with non-zero code.

Follow up question regarding this:
1. so any error that is level 50 (severity: ERROR) or above will always result in renovate exitting?
2. if we pass multiple repositories to renovate, and one of the repository encounter an `ERROR` / `FATAL`, is there any way to still let renovate process the remaining repository (do not exit immediately. Exit with non-zero only after finishing all repositories)

edit: it does not exit early, I saw the wrong log 😞 . It exits at the end after processing all repo with non-zero exit code

**Reply by rarkins** [reply]
Only fatal-level log messages cause Renovate to exit immediately if in the middle of a repo. Error-level log messages often do not cause the whole Renovate to stop. Instead, such messages are gathered and logged _a second time_ right before exit, so that it's easy for you to see all such errors in one place.

**Reply by chfern** [reply]
> Error-level log messages often do not cause the whole Renovate to stop

Does that mean there is a possibility where an ERROR-level log cause renovate to stop immediately?

**Reply by chfern** [reply]
Also, from renovate's point of view, what is classified as WARN, ERROR and FATAL?

Example, given the scenario where renovate fails to process a particular repo, some error type are reported with WARN, some with ERROR
```
// fail due to missing / invalid branch, log level ERROR
{"name":"renovate","hostname":"omitted","pid":18,"level":50,"logContext":"omitted","repository":"username/thirdRepo","err":{"task":{"commands":["rev-parse","HEAD"],"format":"utf-8","parser":"[function]"},"message":"HEAD\nfatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree ...

// fail due to no access, log level WARN
{"name":"renovate","hostname":"omitted","pid":15,"level":40,"logContext":"omitted","repository":"username/firstRepo","branch":"renovate-0-go-github.com/gin-gonic/gin-vulnerability","err":{"task":{"commands":["push","origin","renovate-0-go-github.com/gin-gonic/gin-vulnerability:renovate-0-go-github.com/gin-gonic/gin-vulnerability","--force-with-lease","-u","--no-verify","--verbose","--porcelain"],"format":"utf-8","parser":"[function]"},"message":"Pushing to https://gitlabhost/username/firstRepo.git\nremote: You are not allowed to push code to this project.\nfatal: unable to access 'https://gitlabhost/username/firstRepo.git/': The requested URL returned error: 403\n","stack":"Error: Pushing to https://gitlabhost/username/firstRepo.git\nremote: You are not allowed to push code to this project.\nfatal: unable to access 'https://gitlabhost/username/firstRepo.git/': The requested URL returned error: 403\n\n    at Object.action (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/opt/containerbase/tools/renovate/35.138.1/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"},"msg":"Error updating branch","time":"2023-09-04T08:02:46.698Z","v":0}
```

**Reply by rarkins** [reply]
It doesn't usually happen, and I can't think of any examples of a mid-repo error which causes a global stop, but I wouldn't guarantee it will never happen.

**Reply by rarkins** [reply]
> Also, from renovate's point of view, what is classified as WARN, ERROR and FATAL?

We are still trying to establish some consistency here.

FATAL is easiest - these are problems which mean Renovate can't continue and must exit immediately. Often user error and early in the initialization, but sometimes can be at runtime e.g. due to disk or memory.

We elevate messages to WARN if we think the _users_ should be aware of it. Warn messages result in messages in the dependency dashboard, for example. Things which should not trouble users are kept as info or debug.

ERROR messages at runtime (during a repo) should ideally be for "this shouldn't happen" problems that we as maintainers would like to know about. For example in the hosted app we run, we aim to check and fix every error we find.

**Reply by chfern** [reply]
got it, thankss
